src/WinRT.Runtime/ObjectReference.cs

```diff
@@ -673,8 +673,20 @@ private static unsafe T GetVtable(IntPtr thisPtr)
             // so we instead fall back to typeof(T).IsGenericType, which is not a JIT-time constant.
 #if NET
             if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
-            {
-                throw new NotSupportedException("Managed vtable types (ie. containing any reference types) are not supported.");
+            {
+                if (!RuntimeFeature.IsDynamicCodeCompiled)
+                {
+                    throw new NotSupportedException("Managed vtable types (ie. containing any reference types) are not supported.");
+                }
+
+                return GetVtableForJitEnvironment(thisPtr);
+
+                [UnconditionalSuppressMessage("Trimming", "IL2090", Justification = "Fallback method for JIT environments that is not trim-safe by design.")]
+                [MethodImpl(MethodImplOptions.NoInlining)]
+                static T GetVtableForJitEnvironment(IntPtr thisPtr)

```

Comment:
I'm fine with the change, given it doesn't impact AOT, but technically speaking this introduces undefined behavior for trimming, since we'll now generate no warnings, but code using this path with trimming will break with no warnings. I don't really have a good solution, just pointing this out.